File: app-android/vela-wallet/app/src/main/java/app/getvela/wallet/feature/send/core/SweepPick.kt
package app.getvela.wallet.feature.send.core

import app.getvela.wallet.feature.send.SendLive

/**
 * The sweep pick's shell-side rule (spec 045 D2, desktop 033): the FIRST tick
 * pins the chain (`SetMultiNetwork`) before the toggle; a row on another chain
 * is drawn dimmed and a tap on it is not an event (the core would refuse it —
 * a tappable row is an invitation it will not honour); clearing the last tick
 * unpins, so the next tick pins again. Everything else — which rows are
 * valuable, how much moves — is the core's.
 */
object SweepPick {
    /** The events one tap on `tokenId` becomes, in order; empty when the row is dimmed. */
    fun tap(view: SendView, tokenId: String): List<SendEvent> {
        val token = view.tokens.firstOrNull { SendLive.tokenId(it) == tokenId } ?: return emptyList()
        val pinned = view.multi_chain_id
        val selected = tokenId in view.multi_selected_ids
        return when {
            selected && view.multi_selected_ids.size == 1 ->
                listOf(SendEvent.ToggleMultiToken(tokenId), SendEvent.SetMultiNetwork(null))
            selected -> listOf(SendEvent.ToggleMultiToken(tokenId))
            pinned == null -> listOf(SendEvent.SetMultiNetwork(token.chain_id), SendEvent.ToggleMultiToken(tokenId))
            pinned != token.chain_id -> emptyList()
            else -> listOf(SendEvent.ToggleMultiToken(tokenId))
        }
    }

    /**
     * "Select all valuable" over the rows on screen: with no chain pinned yet,
     * the first visible valuable row's chain is pinned first, so the core's
     * scoped toggle has one chain to work in.
     */
    fun selectAll(view: SendView, visibleIds: List<String>): List<SendEvent> {
        if (view.multi_chain_id != null) return listOf(SendEvent.ToggleAllMultiTokens(visibleIds))
        val first = view.tokens.firstOrNull { SendLive.tokenId(it) in visibleIds && SendLive.tokenId(it) in view.multi_valuable_ids }
            ?: return emptyList()
        return listOf(SendEvent.SetMultiNetwork(first.chain_id), SendEvent.ToggleAllMultiTokens(visibleIds))
    }

    /** Which rows are drawn dimmed: another chain's, once one is pinned. */
    fun dimmed(view: SendView): List<Boolean> =
        view.tokens.map { token -> view.multi_chain_id != null && token.chain_id != view.multi_chain_id }
}

File: app-ios/VelaWallet/VelaWallet/Features/Send/SweepPick.swift
//
//  SweepPick.swift
//  VelaWallet
//
//  What one tap on the sweep picker means.
//
//  Ported from `app-android/.../feature/send/core/SweepPick.kt` (spec 045
//  research D2). Every judgement is the core's — which tokens are worth
//  sweeping, how much of each actually moves — and what is here is only the
//  order of the events one tap becomes.
//
//  ## The picking flag is the shell's, and only that flag
//
//  The core's `multiSelectMode` flips at **confirm**, not when the tick boxes
//  appear, so "are we picking" cannot be read from it. That one flag lives in
//  the screen; every other question goes to the core.
//
//  ## Unticking the last token releases the chain
//
//  Without it, somebody who ticks the wrong token first is pinned to that chain
//  until they leave the screen entirely — the wallet having quietly decided
//  something they did not mean to decide.
//

import Foundation

enum SweepPick {

    /// The events one tap becomes, **in order**. Empty means the row is dimmed
    /// and the tap is not an answer to anything.
    static func tap(view: SendViewWire, tokenId: String, chainId: Int) -> [[String: Any]] {
        let selected = view.multiSelectedIds.contains(tokenId)
        let pinned = view.multiChainId

        if selected, view.multiSelectedIds.count == 1 {
            return [
                ["type": "toggle_multi_token", "token_id": tokenId],
                ["type": "set_multi_network", "chain_id": NSNull()],
            ]
        }
        if selected {
            return [["type": "toggle_multi_token", "token_id": tokenId]]
        }
        if pinned == nil {
            return [
                ["type": "set_multi_network", "chain_id": chainId],
                ["type": "toggle_multi_token", "token_id": tokenId],
            ]
        }
        // A row on another chain is drawn dimmed and is not tappable; this is
        // the second lock on the same door.
        if pinned != chainId { return [] }
        return [["type": "toggle_multi_token", "token_id": tokenId]]
    }

    /// "Select all valuable", scoped to what is **on screen**.
    ///
    /// The shell supplies the visible ids and nothing else: which of them are
    /// valuable is the core's, and a filtered list must not sweep rows a person
    /// cannot see.
    static func selectAll(
        view: SendViewWire, visibleIds: [String], chainOf: (String) -> Int?
    ) -> [[String: Any]] {
        if view.multiChainId != nil {
            return [["type": "toggle_all_multi_tokens", "visible_ids": visibleIds]]
        }
        // Nothing pinned yet: the first VALUABLE visible row decides the chain.
        guard let first = visibleIds.first(where: { view.multiValuableIds.contains($0) }),
              let chainId = chainOf(first)
        else { return [] }
        return [
            ["type": "set_multi_network", "chain_id": chainId],
            ["type": "toggle_all_multi_tokens", "visible_ids": visibleIds],
        ]
    }

    /// Whether a row is on a chain the pick has left behind.
    static func dimmed(view: SendViewWire, chainId: Int) -> Bool {
        guard let pinned = view.multiChainId else { return false }
        return pinned != chainId
    }
}
